### ISSUE TITLE
[feature request] Apollo federation

### ISSUE THREAD

**Issue by remorses** [tartiflette/tartiflette#299]
The spec of Apollo federation allows to build graphql servers that can be composed one with another, i think it would be cool to implement the spec in tartiflette.
The implementation consist in implementing the following types, then the schema can be integrated in an apollo federation gateway
```
scalar _Any
scalar _FieldSet

# a union of all types that use the @key directive
union _Entity

type _Service {
  sdl: String
}

extend type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
}

directive @external on FIELD_DEFINITION
directive @requires(fields: _FieldSet!) on FIELD_DEFINITION
directive @provides(fields: _FieldSet!) on FIELD_DEFINITION
directive @key(fields: _FieldSet!) on OBJECT | INTERFACE

# this is an optional directive discussed below
directive @extends on OBJECT | INTERFACE
```

The spec is [here](https://www.apollographql.com/docs/apollo-server/federation/federation-spec/)
[Here](https://github.com/erebus1/graphene-federation/blob/master/graphene_federation/main.py) is an attempt of implementing it for graphene

**Comment by abusi** [comment]
Hi @remorses, Yep, this could be a cool feature. And it is well suited to be in a Tartiflette-plugin-apollo-federation.
I have to dive in the spec to see if we have everything ready in the engine for it to be implementable. But yes, it's a very good idea.
Do you wanna work on it ?

**Comment by remorses** [comment]
Yes i will work on it. Do you know if there is a way to access the complete sdl string from the resolver arguments? The spec requires that `_Service.sdl` returns the complete graphql schema. Or maybe the schema node classes can be dumped in sdl syntax?

**Comment by abusi** [comment]
Hum, that why I have to take time to read this spec, There are somethings that are not doable rightnow, I need to find what is needed in the Engine itself so you could implemented the Plugin and we can implement the Engine/Schema interface for you to use.

**Comment by remorses** [comment]
I made a basic implementation at https://github.com/remorses/tartiflette-plugin-apollo-federation
Still it needs an implementation of the `@key` directive and surrounding pieces.
For that to work i need `on_bake` method in the directives class.

**Comment by abusi** [comment]
Yeah we'are working on this rigth now at the office. As it is a requirement for a lot of coool features, we wanna do it right, that's why we're taking some time to do it.

I'll ping you on the PR when it will be ready to show and comment. It's coming soon.

**Comment by jugdizh** [comment]
I'm interested in this feature too. Are there any updates?

Thanks

**Comment by abusi** [comment]
@jugdizh the plan is to have this possible in v2. RN @Maximilien-R is working on a refactoring of the validation side of the schema and query which will be the starting point of the v2 branch in which we will completly rework the schema side of tartiflette to enable these kind of feature. I think, we'll be there for 2021.

**Comment by ca-simone-chiorazzo** [comment]
@abusi Do you have an ETA for the V2 version containing this feature? We are evaluating it and we would use something of maintained because used in production. Unfortunately, graphene seems dead.

**Comment by abusi** [comment]
@ca-simone-chiorazzo Sadly, V2 will only contains tools that would help use make this, V2 should happen in 2021, but I can't get you an ETA as I'm not focused on it RN.
Tartiflette is used in production by at least 2 web companies that I know of (Dailymotion being one) without any issues for more than 2 years now, but they do not need federation.

If you really need it, and are interested in tartiflette, feel free to submit any PR, i'll gladly review it.
In the end, only some directive will need to be implemented in order to support it.

**Comment by ca-simone-chiorazzo** [comment]
@abusi Thank you for the clarification!

**Comment by gc-ss** [comment]
> V2 should happen in 2021

Excited to have federation support!

**Comment by abusi** [comment]
@gc-ss Sadly, the development speed of tartiflette is not moving fast. Appolo federation work hasn't started :D, sorry to disapoint, I don't really know if we're gonna do it :/

**Comment by gc-ss** [comment]
I really appreciate providing this feedback!

**Comment by abusi** [comment]
@gc-ss I'm a little bit disconnected from the GQL world, I know we want to implem GQL-2021 (since we already have some of the feature, we don't miss a lot), but is `appolo` still a big thing in the GQL community or does the hype has gone down ?

**Comment by gc-ss** [comment]
I won't say I am by any means connected to the GQL world either but I am looking for a framework that lets me stitch together various independent services (subgraphs) while exposing them through one API Gateway (supergraph)

Whether this is done using apollo's federation solution or something else, is secondary to having any support for it at all.

For example, taking a stab at https://github.com/movio/bramble or https://github.com/gmac/schema-stitching-handbook is a welcome pursuit 👍

**Comment by abusi** [comment]
Hum, tartiflette doesn't support schema stitching, in the sense that tartfilette does need the complete schema to work. But on a complete schema, a resolver can query a remote service without any problem.
